README
NAME
    Test::Ping - Testing pings using Net::Ping

VERSION
    Version 0.18

SYNOPSIS
    This module helps test pings using Net::Ping

        use Test::More tests => 2;
        use Test::Ping;

        my $good_host = '127.0.0.1';
        my $bad_host  = '1.1.1.1;

        ping_ok(     $good_host, "able to ping $good_host" );
        ping_not_ok( $bad_host,  "can't ping $bad_host"    );
        ...

DESCRIPTION
    Using this module you do not have to work with an object, but can
    instead use actual procedural testing functions, which is cleaner and
    more straight forward for testing purposes. This module keeps track of
    the object for you, starting and closing it and provides a nifty way of
    testing for pings.

SUBROUTINES/METHODS
  ping_ok( $host, $test )
    Checks if a host replies to ping correctly.

    This returns the return value and duration, just like Net::Ping's
    "ping()" method.

  ping_not_ok( $host, $test )
    Does the exact opposite of "ping_ok()".

  create_ping_object_ok( @args, $test )
    This tries to create a ping object and reports a fail or success. The
    args that should be sent are whatever args used with Net::Ping.

  create_ping_object_not_ok( @args, $test )
    Tried to create a ping object and attempts to fail. The exactly opposite
    of the "create_ping_object_not_ok()"

EXPORT
    "ping_ok"

    "ping_not_ok"

    "create_ping_object_ok"

    "create_ping_object_not_ok"

SUPPORTED VARIABLES
    Variables in Test::Ping are tied scalars. Some variables change the
    values in the object hash while others run methods. This follows the
    behavior of Net::Ping. Below you will find each support variable and
    what it changes.

  BIND
    Runs the "bind" method.

  PROTO
    Changes the "proto" hash value.

  TIMEOUT
    Changes the "timeout" hash value.

  PORT
    Changes the "port_num" hash value.

  HIRES
    Changes the package variable $hires. By default, it is enabled.

  SOURCE_VERIFY
    Changes the package variable $source_verify.

  SERVICE_CHECK
    Changes the "econnrefused" hash value.

INTERNAL METHODS
  _has_var_ok( $var_name, $var_value, $description )
    Gets a variable name to test, what to test against and the name of the
    test. Runs an actual test using Test::Builder.

    This is used to debug the actual module, if you wanna make sure it
    works.

        use Test::More tests => 1;
        use Test::Ping;

        # Test::Ping calls the protocol variable 'PROTO',
        # but Net::Ping calls it internally (in the hash) 'proto'
        # (this is documented above under PROTO)
        # this is checking against Net::Ping specifically

        $Test::Ping::PROTO = 'icmp';
        Test::Ping::_has_var_ok(
            'proto',
            'icmp',
            'Net::Ping has correct protocol variable',
        );

  _ping_object
    When debugging behavior, fetching an internal object from a procedural
    module can be a bit difficult (especially when it has base inheritance
    with another one).

    This method allows you (or me) to fetch the actual Net::Ping object from
    "Test::Ping". It eases testing and assurance.

    This is used by the Tie functions to set the variables for the object
    for you.

        use Test::Ping;
        use Data::Dumper;

        print 'Object internals: ' . Dumper( Test::Ping->_ping_object() );

    Or you could also change the Net::Ping object to one of your own:

        use Test::Ping;
        use Net::Ping;

        Test::Ping->_ping_object( Net::Ping->new(@opts) );

    And doing it with tests:

        use Test::More tests => 2;
        use Test::Ping;

        create_ping_object_ok( 'tcp', 2, 'Creating our own Net::Ping object' );
        ping_ok( $target, "Yay! We can reach $target" );

    However, you should be warned. I test for a Net::Ping object so trying
    to pass other objects will fail. If anyone needs this changed or any
    reason, contact me and I'll consider it.

DEPENDENCIES
    This module uses Net::Ping, Tie::Scalar and Carp.

    Test::Timer is used in the test suite.

AUTHOR
    Sawyer X, "<xsawyerx at cpan.org>"

BUGS
    Please report any bugs or feature requests to "bug-test-ping at
    rt.cpan.org", or through the web interface at
    <http://rt.cpan.org/NoAuth/ReportBug.html?Queue=Test-Ping>.

    There is also a GitHub issue tracker at
    <http://github.com/xsawyerx/test-ping/issues> which I'll probably check
    just as much.

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Test::Ping

    If you have Git, this is the clone path:

    git@github.com:xsawyerx/test-ping.git

    You can also look for information at:

    *   GitHub Website:

        <http://github.com/xsawyerx/test-ping/tree/master>

    *   RT: CPAN's request tracker

        <http://rt.cpan.org/NoAuth/Bugs.html?Dist=Test-Ping>

    *   AnnoCPAN: Annotated CPAN documentation

        <http://annocpan.org/dist/Test-Ping>

    *   CPAN Ratings

        <http://cpanratings.perl.org/d/Test-Ping>

    *   Search CPAN

        <http://search.cpan.org/dist/Test-Ping/>

ACKNOWLEDGEMENTS
    Steve Bertrand (STEVEB) provided many fixes and improvements. Big thank
    you for all the work done.

    Thanks to everyone who works and contributed to "Net::Ping". This module
    depends solely on it.

COPYRIGHT & LICENSE
    Copyright 2009-2010 Sawyer X, all rights reserved.

    This program is free software; you can redistribute it and/or modify it
    under the terms of either:

    *   the GNU General Public License as published by the Free Software
        Foundation; either version 2, or (at your option) any later version,
        or

    *   the Artistic License version 2.0.



Changes
Revision history for Test-Ping

{{$NEXT}}

0.210     2020-06-07 09:41:16+03:00 Asia/Jerusalem

        - A set of fixed by Steve Bertrand:
        - Added prereq Time::HiRes v2.41.
        - Changed RFC 1918 address 172.29.249.249 to doc address 203.0.113.90
          in t/08, t/11 and t/12, as the 1918 address could legitimately be
          in use.
        - Removed 127.0.01 from t/12, as it was failing on some Testers.

0.204     2019-10-31 12:18:43+02:00 Asia/Jerusalem

        - Some cleanups.

0.203     2016-03-13 22:18:34+01:00 Europe/Amsterdam
        - Ugh. $VERSION wasn't added and the indexing failed.
        - Fix the LICENSING.
        - Fix the versioning.
        - Update the PODs automatically via dzil.

0.202   13.03.16
        - Previous release had no Changes entry. Bah.

0.201   13.03.16
        - Dzilify.

0.200   13.03.16
        - fixed typo in is_admin_user() definition which was breaking tests
          on Windows (Undefined subroutine &main::is_admin_user called at
          t\03-icmp.t line 2)
        - fixes RT #108683; ftp.geocities.com unavailable broke t/08-ping_tcp.t
          and added more current/reliable domains to test against
        - Net::Ping, in v2.40, changed to using HiRes by default, fixed
          t/09-ping_hires.t to conform to said change. Bumped prereq ver of
          Net::Ping to 2.41, to ensure future test compatibility
        - fixed issue in create_ping_object_not_ok(), where it didn't appear
          that it could possibly do the right thing
        - fixed issue in create_ping_object_ok() where the else() statement
          would never have been triggered
        - skip test "Failed test 'IP should still be reachable' at
          t\13-service.t line 133." until the issue can be found/rectified
          (problem only exists in Windows)
        - fixed RELEASE_TESTING test files so they simply skip if the required
          modules aren't installed (critic.t, meta.t and pod.t)
        - removed t/minimumversion.t, it wasn't being used anyhow
        - added configure_requires directive in Build.PL
        - added t/16-ties.t, tests all Ties
        - added t/17-ping_obj_not_ok.t
        - added t/18-ping_obj_ok.t
        - major updates to MAKEFILE.SKIP
        - added t/18-ping_obj_ok.t
        - test coverage now 96.6% from 86.16%

0.17    06.06.10
        fixing Build.PL executability (turned it off)
        add examples
        wrapped POD lines

0.16    15.02.10
        requiring environment variable for test that keeps failing

0.15    12.02.10
        removed boilerplate test
        as god is my non-existent witness, i will pass the tests!

0.14    11.02.10
        minor changes to pass tests

0.13    10.02.10
        repository in META.yml using Module::Build (cpanservice)
        allowng a longer time gap for a test
        correcting Changelog

0.12    13.08.09
        repository in META.yml using Module::Build (cpanservice)
        allowng a longer time gap for a test

0.11    02.08.09
        added author tests using Test::XT (thanks Adam Kennedy)
        added Makefile.PL and META.yml instead of leaving it to ./Build dist

0.10    02.08.09
        added explicit terms as GPLv2 or later, or artistic
        (thanks Andy Lester, Perlbuzz.com)

0.09    29.04.09
        ported 300_ping_stream.t to t/10-ping_stream.t
        ported 400_ping_syn.t to t/11-ping_syn.t
        ported 410_syn_host.t to t/12-syn_host.t
        ported 450_service.t to t/13-service.t
        ported 500_ping_icmp.t to t/14-ping_icmp.t
        ported 510_ping_udp.t to t/15-ping_udp.t
        (finished porting all tests from Net::Ping)
        added SERVICE_CHECK, SOURCE_VERIFY, BIND
        corrected POD _ping_object() example
        cleaned up all Ties
        added Carp to Build.PL requires

0.08    28.04.09
        fixed t/03-icmp.t test that prevented installation when admin tests
        added method (and POD for) create_ping_object_ok()
        fixed t/05-udp_tcp_stream_syn_inst.t to use create_ping_object_ok()
        ping_ok() now returns return value and duration like Net::Ping's ping()
        added HIRES support even though I thought I wouldn't
        ported HIRES test from Net::Ping (t/09-ping_hires.t)

0.07    27.04.09
        rewrote t/06-alarm.t
        rewrote Test::Ping::Ties::TIMEOUT, Test::Ping::Ties::TIMEOUT
        added Test::Timer as possible skip for testing of t/06-alarm.t
        added Test::Ping::Ties::PORT
        added t/08-ping_tcp.t
        fixed no. of tests in POD synopsis of Ping.pm
        added a lot to the POD

0.06    26.04.09
        moved t/05-udp_tcp_inst.t to t/05-udp_tcp_stream_syn_inst.t
        adding test for stream inst and syn inst
        added t/06-alarm.t
        added ping_not_ok()

0.05    26.04.09
        some POD changes and subroutine returns to keep up with perlcritic
        added -T to tests
        added 03-icmp.t
        added 04-object.t and _ping_object()
        introducing Tie::Scalar to replace _update_variables()
        added Test::Ping::Ties::PROTO.pm
        removed t/02-variables.t
        updated POD
        removed pos_bug_net_ping.t - it's not a bug

0.04    24.04.09
        added test for possible bug in Net::Ping
        got it stable
        removed BIND and PORT support for now

0.03    22.04.09
        separated variable tests to a different file
        added TIMEOUT, PORT, BIND, SOURCE_VERIFY, SERVICE_CHECK,
            TCP_SERVICE_CHECK

0.02    22.04.09
        ping_ok() now works

0.01    22.04.09
        Basic skeleton



lib/Test/Ping.pm
package Test::Ping;
# ABSTRACT: Testing pings using Net::Ping

use strict;
use warnings;

use Test::Ping::Ties::BIND;
use Test::Ping::Ties::PORT;
use Test::Ping::Ties::PROTO;
use Test::Ping::Ties::HIRES;
use Test::Ping::Ties::TIMEOUT;
use Test::Ping::Ties::SOURCE_VERIFY;
use Test::Ping::Ties::SERVICE_CHECK;

my  $CLASS         = __PACKAGE__;
my  $OBJPATH       = __PACKAGE__->builder->{'_net-ping_object'};
my  $method_ignore = '__NONE';
our @EXPORT        = qw(
    ping_ok
    ping_not_ok
    create_ping_object_ok
    create_ping_object_not_ok
);

# Net::Ping variables
our $PORT;
our $BIND;
our $PROTO;
our $HIRES;
our $TIMEOUT;
our $SOURCE_VERIFY;
our $SERVICE_CHECK;

BEGIN {
    use parent 'Test::Builder::Module';
    use Net::Ping;

    __PACKAGE__->builder->{'_net-ping_object'} = Net::Ping->new($PROTO);

    ## no critic qw(Miscellanea::ProhibitTies)
    tie $PORT,          'Test::Ping::Ties::PORT';
    tie $BIND,          'Test::Ping::Ties::BIND';
    tie $PROTO,         'Test::Ping::Ties::PROTO';
    tie $HIRES,         'Test::Ping::Ties::HIRES';
    tie $TIMEOUT,       'Test::Ping::Ties::TIMEOUT';
    tie $SOURCE_VERIFY, 'Test::Ping::Ties::SOURCE_VERIFY';
    tie $SERVICE_CHECK, 'Test::Ping::Ties::SERVICE_CHECK';
}

sub ping_ok {
    my ( $host, $name ) = @_;
    my $tb     = $CLASS->builder;
    my $pinger = $OBJPATH;

    my ( $ret, $duration ) = $pinger->ping( $host, $TIMEOUT );
    $tb->ok( $ret, $name );

    return ( $ret, $duration );
}

sub ping_not_ok {
    my ( $host, $name ) = @_;
    my $tb     = $CLASS->builder;
    my $pinger = $OBJPATH;

    my $alive = $pinger->ping( $host, $TIMEOUT );
    $tb->ok( !$alive, $name );

    return 1;
}

sub create_ping_object_ok {
    my @args = @_;
    my $name = pop @args || q{};
    my $tb   = $CLASS->builder;

    my $success = eval { $OBJPATH = Net::Ping->new(@args); 1; };

    $tb->ok( $success && ref $OBJPATH eq 'Net::Ping', $name );
}

sub create_ping_object_not_ok {
    my @args = @_;
    my $name = pop @args || q{};
    my $tb   = $CLASS->builder;
    my $error;
    eval { Net::Ping->new(@args); 1; }
    or $error = $@;

    $tb->ok( $error, $name );
}

sub _has_var_ok { ## no critic qw( Subroutines::ProhibitUnusedPrivateSubroutines )
    my ( $var_name, $var_value, $name ) = @_;
    my $tb = $CLASS->builder;
    $tb->is_eq( $OBJPATH->{$var_name}, $var_value, $name );
    return 1;
}

sub _ping_object { ## no critic qw( Subroutines::ProhibitUnusedPrivateSubroutines Subroutines::RequireArgUnpacking )
    my $obj = $_[1] || $_[0] || q{};

    if ( ref $obj eq 'Net::Ping' ) {
        $OBJPATH = $obj;
    }

    return $OBJPATH;
}

END { $OBJPATH->close(); }

1;

__END__

=head1 SYNOPSIS

This module helps test pings using Net::Ping

    use Test::More tests => 2;
    use Test::Ping;

    my $good_host = '127.0.0.1';
    my $bad_host  = '1.1.1.1;

    ping_ok(     $good_host, "able to ping $good_host" );
    ping_not_ok( $bad_host,  "can't ping $bad_host"    );
    ...

=head1 DESCRIPTION

Using this module you do not have to work with an object, but can instead use
actual procedural testing functions, which is cleaner and more straight forward
for testing purposes. This module keeps track of the object for you, starting
and closing it and provides a nifty way of testing for pings.

=head1 SUBROUTINES/METHODS

=head2 ping_ok( $host, $test )

Checks if a host replies to ping correctly.

This returns the return value and duration, just like L<Net::Ping>'s C<ping()>
method.

=head2 ping_not_ok( $host, $test )

Does the exact opposite of C<ping_ok()>.

=head2 create_ping_object_ok( @args, $test )

This tries to create a ping object and reports a fail or success. The args that
should be sent are whatever args used with L<Net::Ping>.

=head2 create_ping_object_not_ok( @args, $test )

Tried to create a ping object and attempts to fail. The exactly opposite of the
C<create_ping_object_not_ok()>

=head1 EXPORT

C<ping_ok>

C<ping_not_ok>

C<create_ping_object_ok>

C<create_ping_object_not_ok>

=head1 SUPPORTED VARIABLES

Variables in L<Test::Ping> are tied scalars. Some variables change the values
in the object hash while others run methods. This follows the behavior of
L<Net::Ping>. Below you will find each support variable and what it changes.

=head2 BIND

Runs the C<bind> method.

=head2 PROTO

Changes the C<proto> hash value.

=head2 TIMEOUT

Changes the C<timeout> hash value.

=head2 PORT

Changes the C<port_num> hash value.

=head2 HIRES

Changes the package variable C<$hires>. By default, it is enabled.

=head2 SOURCE_VERIFY

Changes the package variable C<$source_verify>.

=head2 SERVICE_CHECK

Changes the C<econnrefused> hash value.

=head1 INTERNAL METHODS

=head2 _has_var_ok( $var_name, $var_value, $description )

Gets a variable name to test, what to test against and the name of the test.
Runs an actual test using L<Test::Builder>.

This is used to debug the actual module, if you wanna make sure it works.

    use Test::More tests => 1;
    use Test::Ping;

    # Test::Ping calls the protocol variable 'PROTO',
    # but Net::Ping calls it internally (in the hash) 'proto'
    # (this is documented above under PROTO)
    # this is checking against Net::Ping specifically

    $Test::Ping::PROTO = 'icmp';
    Test::Ping::_has_var_ok(
        'proto',
        'icmp',
        'Net::Ping has correct protocol variable',
    );

=head2 _ping_object

When debugging behavior, fetching an internal object from a procedural module
can be a bit difficult (especially when it has base inheritance with another
one).

This method allows you (or me) to fetch the actual L<Net::Ping> object from
C<Test::Ping>. It eases testing and assurance.

This is used by the Tie functions to set the variables for the object for you.

    use Test::Ping;
    use Data::Dumper;

    print 'Object internals: ' . Dumper( Test::Ping->_ping_object() );

Or you could also change the Net::Ping object to one of your own:

    use Test::Ping;
    use Net::Ping;

    Test::Ping->_ping_object( Net::Ping->new(@opts) );

And doing it with tests:

    use Test::More tests => 2;
    use Test::Ping;

    create_ping_object_ok( 'tcp', 2, 'Creating our own Net::Ping object' );
    ping_ok( $target, "Yay! We can reach $target" );

However, you should be warned. I test for a L<Net::Ping> object so trying to
pass other objects will fail. If anyone needs this changed or any reason,
contact me and I'll consider it.

=head1 DEPENDENCIES

This module uses L<Net::Ping>, L<Tie::Scalar> and L<Carp>.

L<Test::Timer> is used in the test suite.

=head1 BUGS

Please report any bugs or feature requests on the a GitHub issue tracker at
L<http://github.com/xsawyerx/test-ping/issues>.

=head1 ACKNOWLEDGEMENTS

Steve Bertrand (STEVEB) provided many fixes and improvements. Big thank
you for all the work done.

Thanks to everyone who works and contributed to C<Net::Ping>. This module
depends solely on it.


lib/Test/Ping/Ties/BIND.pm
package Test::Ping::Ties::BIND;
# ABSTRACT: Bind Tie variable to Test::Ping

use strict;
use warnings;

use Net::Ping;
use Test::Ping;
use Tie::Scalar;
use Carp;

sub TIESCALAR { return bless {}, shift;                    }
sub FETCH     { carp 'Usage: $p->bind($local_addr)';       }
sub STORE     { Test::Ping->_ping_object()->bind( $_[1] ); }

1;

__END__

=head1 DESCRIPTION

In order to allow complete procedural interface to Net::Ping, even though it's
an object, I use a Tie::Scalar interface. Every variable is also defined
separately to make it cleaner and easier.

At some point they might be joined together in a single file, but I doubt it.

Please refrain from using this directly.

=head1 EXPORT

None.


lib/Test/Ping/Ties/PORT.pm
package Test::Ping::Ties::PORT;
# ABSTRACT: Port Tie variable to Test::Ping

use strict;
use warnings;

use Net::Ping;
use Test::Ping;
use Tie::Scalar;

sub TIESCALAR { return bless {}, shift;                           }
sub FETCH     { return Test::Ping->_ping_object()->{'port_num'};  }
sub STORE     { Test::Ping->_ping_object()->{'port_num'} = $_[1]; }

1;

__END__

=head1 DESCRIPTION

In order to allow complete procedural interface to Net::Ping, even though it's
an object, I use a Tie::Scalar interface. Every variable is also defined
separately to make it cleaner and easier.

At some point they might be joined together in a single file, but I doubt it.

Please refrain from using this directly.

=head1 EXPORT

None.


Build.PL
use strict;
use warnings;
use Module::Build;

my $builder = Module::Build->new(
    meta_merge => {
        resources => {
            repository => 'http://github.com/xsawyerx/test-ping',
        },
    },

    module_name         => 'Test::Ping',
    license             => 'perl',
    dist_author         => 'Sawyer X <xsawyerx@cpan.org>',
    dist_version_from   => 'lib/Test/Ping.pm',
    build_requires => {
        'Test::More'  => 0,
        'Test::Timer' => 0.05,
    },

    configure_requires => {
        'Module::Build' => 0,
    },
    requires => {
        'Carp'          => 0,
        'Net::Ping'     => 2.41,
        'Tie::Scalar'   => 0,
        'Test::Builder' => 0.86,
    },

    add_to_cleanup      => [ 'Test-Ping-*' ],
    create_makefile_pl  => 'traditional',
    create_readme       => 1,
    create_license      => 1,
);

$builder->create_build_script();
